File: python/dsa/algorithms/dp/2d/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n, m = len(word1), len(word2)
        dp = [[0] * (m + 1) for _ in range(n + 1)] # @viz: init
        
        for i in range(n + 1): dp[i][0] = i
        for j in range(m + 1): dp[0][j] = j
        
        for i in range(1, n + 1): # @viz: loopI
            for j in range(1, m + 1): # @viz: loopJ
                if word1[i-1] == word2[j-1]: # @viz: ifMatch
                    dp[i][j] = dp[i-1][j-1] # @viz: match
                else:
                    dp[i][j] = 1 + min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) # @viz: mismatch
                    
        return dp[n][m] # @viz: result

File: java/dsa/algorithms/dp/2d/edit-distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int n = word1.length(), m = word2.length();
        int[][] dp = new int[n + 1][m + 1]; // @viz: init
        
        for (int i = 0; i <= n; i++) dp[i][0] = i;
        for (int j = 0; j <= m; j++) dp[0][j] = j;

        for (int i = 1; i <= n; i++) { // @viz: loopI
            for (int j = 1; j <= m; j++) { // @viz: loopJ
                if (word1.charAt(i-1) == word2.charAt(j-1)) { // @viz: ifMatch
                    dp[i][j] = dp[i-1][j-1]; // @viz: match
                } else {
                    dp[i][j] = 1 + Math.min(dp[i-1][j-1], Math.min(dp[i-1][j], dp[i][j-1])); // @viz: mismatch
                }
            }
        }
        return dp[n][m]; // @viz: result
    }
}